Lotto game PR 요청합니다.

Egenieee opened the pull request:
아직 MVC 패턴을 정확이 어떤 식으로 적용해야 하는 지 잘 모르겠지만, 할 수 있는 만큼 구현해봤습니다! 구현 하면서 계속 고민 했던 부분이 **Controller 와 Model의 경계** 라고 할까요,,? Controller가 하는 일이 어디까지 커버를 해야하는 지 감이 제일 안오더라구요,, 

제가 알기론 Controller 은 가운데서 값을 받아서 넘겨주는 역할만 하는 걸로 알고 있는데, 제가 짠 코드는 대부분의 로직이 Controller에 있어서 이걸 어떻게 Domain으로 옮겨야 하는 지 잘 모르겠습니다 😅 

Controller에서 view나 다른 Lotto, WinningLotto 객체를 다 사용하긴 했는데 이렇게 하는 게 아닌 거 같아요 하하..
일단 그런 부분은 피드백을 받아보고 싶기도 하고 더 이상 refactoring을 어떻게 해야할 지 모르겠어서 여기서 스탑하고 pr요청합니다..!

바쁘실텐데 감사합니다 😭

Egenieee committed:
docs: 구현 기능 README.md에 추가

Egenieee committed:
MVC package 생성
- docs: 구현 기능 README.md에 추가
- feat: inputView 이용해서 Lotto 구매 금액 입력 받는 기능 추가

Egenieee committed:
docs: README.md 수정

Egenieee committed:
OutputView 이용하여 몇 개의 로또를 구매했는 지 출력 기능 추가

Egenieee committed:
구매 금액 만큼의 로또 객체 생성하는 기능 추가
- feat: 랜덤으로 로또 번호를 생성하는 CreateLottoNumber class 추가
- feat: 로또 구매 개수 만큼 로또 객체 생성하는 기능 추가

Egenieee committed:
feat: 구매한 로또 OutputView 이용하여 출력하는 기능 추가

Egenieee committed:
docs: README.md 구현 기능 수정

Egenieee committed:
feat: 지난 주 당첨 번호 입력 받는 기능 추가

Egenieee committed:
feat: 보너스 볼 입력받는 기능 추가

Egenieee committed:
feat: 당첨 로또 번호 List로 split하는 기능 추가

Egenieee committed:
feat: 입력받은 당첨 번호와 보너스 번호로 WinningLotto 객체 생성 기능 추가

Egenieee committed:
fix: winningLotto 객체 생성 기능 수정
- 중복된 숫자를 당첨 번호로 입력할 경우 ERROR 메세지를 띄우고 다시 입력 받도록 수정

Egenieee committed:
feat: 지난 당첨 로또 번호와 생성된 로또가 몇개 맞는 지 확인하는 기능 추가

Egenieee committed:
당첨 통계 기능 추가
- feat: 당첨 통계 계산하는 기능 추가
- feat: 당첨 통계 결과 OutputView로 출력하는 기능 추가

Egenieee committed:
refactor: Controller 수정

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,169 @@
+
+            isValid = checkDuplicatedNumber(winningNumberForCheck);
+
+            InputView.checkValidWinningLottoNumber(isValid);
```
이 로직은 입력을 받는게 아니고 검증만 하고 있어서 View영역이 아니라 다른곳에 있어도 괜찮지않을까?

gch01410 reviewed (changes requested):
잘했는데?? ㅋㅋㅋ
우선 MVC 패턴 관련해서만 리뷰남겼어!
몇개 안되니까 확인해보고! ㅋㅋ
테스트 코드도 작성해서 다시 요청해줘

Egenieee committed:
Domain 영역에서 출력 로직 제거
- refactor: OutputView 수정
- refactor: Lotto 모델에서 출력 로직 제거, getter 생성

Egenieee committed:
잘못된 로또 번호 받을 경우 에러 메세지 출력 로직 수정
- refactor: InputView -> OutputView 에서 출력되도록 수정

Egenieee committed:
Gradle setting & Create  Domain Test Class

Egenieee committed:
feat: CreateLottoNumberTest 작성

Egenieee committed:
feat: LottoTest 작성

Egenieee committed:
WinningLottoTest 작성
-feat: WinningLottoTest 작성
-refactor: Test code 작성 위해서 WinningLotto에 getter 추가

Egenieee committed:
feat: RankTest 작성

Egenieee committed:
ControllerTest 작성
-feat: ControllerTest 작성
-refactor: Test code 작성 위해서 몇 개 메서드 접근제어자 public으로 변경

gch01410 reviewed (changes requested):
테스트 코드 꼼꼼하게 작성 잘했어! 👍 
다만 컨트롤러가 많은 역할을 하고 있는것 같네
컨트롤러는 model영역과 view영역 사이에서 데이터를 전달해주는 역할만 하도록 변경해봐! 
게임진행은 model영역에서 담당하고 게임진행에 필요한 값을 컨트롤러를 통해 view에서 전달도록 구현하면 될 것 같아

Egenieee committed:
Controller 수정
-refactor: Controller의 비즈니스 로직 Model영역으로 옮김.
-refactor: Model영역에 GameHelper class를 추가해 비즈니스 로직을 구현
-refactor: Outputview 수정

Egenieee committed:
Test Class refactor
-refactor: LottoTest 수정
-refactor: CreateLottoNumberTest 수정
-refactor: RankTest 수정
-refactor: GameHelperTest 생성
-refactor: WinningLottoTest 수정

Egenieee committed:
-refactor: Application 변수명 수정

Egenieee commented:
최대한 Controller는 전달자 역할만 하도록 수정해보았는데, 이렇게 되면 막상 문제가 Controller에 대한 Test 코드 작성을 어떤 식으로 해야할 지 모르겠더라구요,,! 아무리 생각해봐도 모르겠어서 일단은 피드백 받은 부분 수정해보고 푸쉬합니다!

gch01410 commented on README.md:
```diff
@@ -96,42 +101,42 @@ public class Application {
-### 프로그래밍 요구사항 - WinningLotto
-- WinningLotto 객체를 활용해 구현해야 한다.
+### 프로그래밍 요구사항 - domain.WinningLotto
+- domain.WinningLotto 객체를 활용해 구현해야 한다.
 - match() 메서드의 반환 값인 Rank는 저장소에서 제공한다.
-- WinningLotto 기본 생성자를 추가할 수 없다.
+- domain.WinningLotto 기본 생성자를 추가할 수 없다.
```
이름 변경하면서 같이 변경된것 같네😄

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,86 @@
+        // 몇 개의 로또를 구매했는 지 출력
+        OutputView.printLottoCount(lottoCount);
+
+        List<Lotto> lottoes = new ArrayList<>();
+
+        helper.createLottoes(lottoes, lottoCount);
```
컨트롤러에서 선언과 초기화를 한 이유가 있을까??

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,86 @@
+
+    // 랜덤 로또 받아서 출력
+    private List<Lotto> getLottoes() {
+        domain.GameHelper helper = new domain.GameHelper();
```
GameHelper클래스를 인스턴스화해서 사용한 이유가 있을까??

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,86 @@
+        List<Integer> winningResult = helper.getLottoResult(winningLotto, lottoes);
+
+        // 당첨 금액과 구매 금액으로 수익률 구하기
+        int totalGetMoney = helper.getLottoTotalMoney(winningLotto, lottoes);
+        double yield = helper.getYield(totalGetMoney, lottoes.size() * LOTTO_PRICE_PER_ONE);
```
수익률을 구하는 과정에서 총 상금을 계산해서 바로 사용해도 되지않을까?

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,86 @@
+    private void printResult(WinningLotto winningLotto, List<Lotto> lottoes) {
+        domain.GameHelper helper = new domain.GameHelper();
+
+        // 당첨 결과 리스트로 받기 순서대로 [3개 일치, 4개일치, 5개 일치, 5개 + 보너스 일치, 6개 일치 개수]
+        List<Integer> winningResult = helper.getLottoResult(winningLotto, lottoes);
```
당첨결과를 List의 순서에 의존하지않고 받아보도록 할 수 있을까?

gch01410 commented on src/main/java/domain/CreateLottoNumber.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.*;
+
+public class CreateLottoNumber {
```
네이밍으로 `LottoNumberCreator` 가 더 적합해보여
클래스 이름은 보통 명사로 구성해!

gch01410 commented on src/main/java/domain/GameHelper.java:
```diff
@@ -0,0 +1,114 @@
+
+    public boolean checkDuplicatedNumber(String winningNumbers) {
+        Set<Integer> splitWinningNumbers = splitWinnigLottoNumber(winningNumbers);
+        return splitWinningNumbers.size() == 6;
```
6이 어떤 숫자인지 의미를 알기 어려울 수 있어
상수로 선언해서 이름으로 의미를 드러내주는게 좋아

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,86 @@
+        List<Integer> winningResult = helper.getLottoResult(winningLotto, lottoes);
+
+        // 당첨 금액과 구매 금액으로 수익률 구하기
+        int totalGetMoney = helper.getLottoTotalMoney(winningLotto, lottoes);
```
당첨금액을 구하는 과정에서 다시 당첨로또를 구하고 있어!
전에 구했던 당첨 결과를 이용해볼 수 있지않을까?

gch01410 commented on src/main/java/domain/GameHelper.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    // 랜덤 숫자로 이루어진 하나의 로또 객체 생성하기.
+    public Lotto createOneLotto() {
```
GameHelper의 메서드에서만 사용하고 있어서 private으로 선언하는게 맞는 것 같아!

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,86 @@
+    private static final int LOTTO_PRICE_PER_ONE = 1000;
+
+    public void run() {
+        List<Lotto> lottoes = getLottoes();
```
Lotto 리스트를 일급컬렉션으로 만들어볼까?
https://jojoldu.tistory.com/412

Egenieee committed:
-refactor: Gamehelper Class static member로 변경

Egenieee committed:
-refactor: Gamehelper Class에서 Lotto List를 return하도록 수정

Egenieee committed:
-refactor: LottoNumberCreator로 Class name 수정

Egenieee committed:
-refactor: 수익률을 구하는 메서드에서 getLottoTotalMoney 메서드를 호출하여 미리 구해 놓은 Map을 통해 상금을 구할 수 있도록 수정

Egenieee committed:
당첨결과를 List의 순서에 의존하지않고 받아보도록 수정
-refactor: List 자료구조로 변경하지 않고 Map을 그대로 사용하도록 GameHelper Class 수정
-refactor: OutputView에서 Map을 넘겨받아 출력할 수 있도록 수정
-refactor: Controller Class 수정

Egenieee committed:
-refactor: GameHelper 메서드 순서 수정

Egenieee committed:
-refactor: GameHelper getLottoTotalMoney 메서드 접근제한자 수정

Egenieee committed:
-refactor: GameHelperTest 수정

Egenieee committed:
Lotto List를 일급컬렉션으로 변경
-feat: Lottoes Class 생성
-refactor: Controller Class 수정
-refactor: GameHelper Class 수정
-refactor: OutputView Class 수정
-refactor: GameHelperTest Class 수정

Egenieee commented:
일급컬렉션을 써보는 건 처음이라 이렇게 하는게 맞는지 잘 모르겠지만,,^^ 일단 피드백 다 반영해서 수정하고 푸쉬합니다!

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,75 @@
+        Map<Rank, Integer> lottoWinResult = GameHelper.getLottoWinResult(winningLotto, lottoes);
+
+        // 수익률 구하기
+        double yield = GameHelper.getYield(lottoWinResult, lottoes.getLottoes().size() * LOTTO_PRICE_PER_ONE);
```
Lottoes가 로또 리스트를 갖고 있으니까 Lottoes가 로또개수를 반환할 수 있지않을까?

gch01410 commented on src/main/java/domain/GameHelper.java:
```diff
@@ -0,0 +1,107 @@
+        // 맞은 로또 결과 저장할 Map 초기화
+        Map<Rank, Integer> lottoResult = createResult();
+
+        for(Lotto lotto : lottoes.getLottoes()) {
+            // 로또 하나 당 당첨 로또랑 비교해서 몇 개 맞았는 지 랭크 확인
+            Rank rank = winningLotto.match(lotto);
+
+            lottoResult.put(rank, lottoResult.getOrDefault(rank, 0) + 1);
```
Lottes가 로또 리스트를 갖고 있으니 WinningLotto를 넘겨주면 Lttoes가 랭크를 구할 수 있지않을까?

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,75 @@
+    // 모델에게 값 받아 로또 당첨 결과 출력
+    private void printResult(WinningLotto winningLotto, Lottoes lottoes) {
+        // 당첨 결과 구하기
+        Map<Rank, Integer> lottoWinResult = GameHelper.getLottoWinResult(winningLotto, lottoes);
```
당첨결과도 일급컬렉션으로 만들어보자!

gch01410 commented on src/main/java/controller/Controller.java:
```diff
@@ -0,0 +1,75 @@
+        return lottoes;
+    }
+
+    // 당첨 로또 받기
```
주석이 없어도 메서드 이름이나 변수명, 클래스 이름 등등.. 으로 로직을 이해할 수 있는게 좋아!
주석은 반드시 필요한 경우에만 사용해줘!

gch01410 reviewed (changes requested):
조금만 더 수정해보자!

Egenieee committed:
-refactor: 일급컬렉션 Lottoes가 로또의 개수 반환할 수 있도록 메서드 추가

Egenieee committed:
-refactor: Lottoes안에서 당첨 결과를 저장할 수 있도록 메서드 추가

Egenieee committed:
-refactor: 불필요한 주석 삭제 처리

Egenieee committed:
-refactor: 사용하지 않는 메서드 삭제

Egenieee committed:
로또 당첨 결과를 일급컬렉션으로 변경
-feat: Lottoes Class 생성
-refactor: Controller Class 수정
-refactor: GameHelper Class 수정
-refactor: OutputView Class 수정
-refactor: Lottoes Class 수정

Egenieee committed:
-refactor: GameHelperTest Class 수정

Egenieee committed:
-feat: LottoesTest Class 작성

Egenieee committed:
-feat: WinningLottoTest Class 작성

Egenieee committed:
-refactor: LottoTest Class 수정

gch01410 commented on src/main/java/domain/GameHelper.java:
```diff
@@ -0,0 +1,91 @@
+        return winningLottoNumber;
+    }
+
+    public static WinningResult getLottoWinResult(WinningLotto winningLotto, Lottoes lottoes) {
+        Map<Rank, Integer> lottoResult = createResult();
+
+        return lottoes.getLottoWinResult(lottoResult, winningLotto);
```
Lottes안에서 lottoResult 맵을 초기화해도 되지않을까?

gch01410 commented on src/main/java/domain/GameHelper.java:
```diff
@@ -0,0 +1,91 @@
+        return (double)totalGetMoney / lottoPrice;
+    }
+
+    private static int getLottoTotalMoney(WinningResult lottoWinResult) {
+        int totalWinMoney = 0;
+
+        for(Rank rank : lottoWinResult.getLottoResult().keySet()) {
+            if(lottoWinResult.getLottoResult().get(rank) > 0) {
+                totalWinMoney += rank.getWinningMoney();
+            }
+        }
+
+        return totalWinMoney;
+    }
```
WinningResult가 로또결과를 알고 있으니 WinningResult가 총합을 구할 수 있지않을까?

gch01410 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,54 @@
+
+    public static void printLottoResult(WinningResult lottoWinResult) {
+        System.out.println();
+        System.out.println(LOTTO_RESULT_OUTPUT_MESSAGE);
+        System.out.printf(LOTTO_FIFTH_OUTPUT_MESSAGE, lottoWinResult.getLottoResult().get(Rank.FIFTH));
+        System.out.printf(LOTTO_FOURTH_OUTPUT_MESSAGE, lottoWinResult.getLottoResult().get(Rank.FOURTH));
+        System.out.printf(LOTTO_THIRD_OUTPUT_MESSAGE, lottoWinResult.getLottoResult().get(Rank.THIRD));
+        System.out.printf(LOTTO_SECOND_OUTPUT_MESSAGE, lottoWinResult.getLottoResult().get(Rank.SECOND));
+        System.out.printf(LOTTO_FIRST_OUTPUT_MESSAGE, lottoWinResult.getLottoResult().get(Rank.FIRST));
```
중복을 제거해보자!
근데 WinningResult의 맵을 꺼내지않고도 값을 구할 수 있지않을까?

gch01410 reviewed (changes requested):
일급컬렉션이 값을 갖고 있기때문에 컬렉션 값을 이용하는 로직을 일급컬렉션 안으로 넣을 수 있어!
일급컬렉션을 사용하면 좋은점에 대해서 알아보자!

Egenieee committed:
-refactor: Lottoes class 안에서 Map을 초기화하고 당첨 결과를 구할 수 있도록 수정

Egenieee committed:
-refactor: WinningResult class 안에서 총 당첨 금액을 구할 수 있도록 수정

Egenieee committed:
-refactor: 오타 수정

Egenieee committed:
OutputView에서 맵 꺼내지 않도록 수정
-feat: WinningResult에 값 넘겨주는 메서드 추가
-refactor: OutputView 수정

Egenieee committed:
-refactor: WinningResultTest 테스트 추가

gch01410 reviewed (approved):
수고했어!

Egenieee committed:
Egenieee
